File: src/3Sum/Solution.py
class Solution:
    """
    @param numbersbers : Give an array numbersbers of n integer
    @return : Find all unique triplets in the array which gives the sum of zero.
    """
    def threeSum(self, numbers):
        # write your code here
        res = []
        numbers.sort()
        
        if len(numbers) < 3:
            return res
        
        for i in range(len(numbers)-2):
            if i > 0 and numbers[i] == numbers[i-1]:
                continue
            j = i + 1
            k = len(numbers) - 1
            while j < k:
                sum = numbers[i] + numbers[j] + numbers[k]
                if sum == 0 and [numbers[i], numbers[j], numbers[k]] not in res:
                    res.append([numbers[i], numbers[j], numbers[k]])
                    j += 1
                    k -= 1
                elif sum < 0:
                    j += 1
                else:
                    k -= 1
        
        return res

File: src/3Sum/Solution.cpp
class Solution {
public:    
    /**
     * @param numbers : Give an array numbers of n integer
     * @return : Find all unique triplets in the array which gives the sum of zero.
     */
    vector<vector<int> > threeSum(vector<int> &nums) {
        // write your code here
        vector<vector<int>> res;
        
        if(nums.size() < 3){
            return res;
        }
        
        sort(nums.begin(), nums.end());
        
        for(int i = 0; i < nums.size() - 2; i++){
            if(i > 0 && nums[i] == nums[i-1]){
                continue;
            }
            int j = i + 1;
            int k = nums.size() - 1;
            while(j < k){
                if(j > i+1 && nums[j] == nums[j-1]){
                    j++;
                    continue;
                }
                int sum = nums[i] + nums[j] + nums[k];
                if(sum == 0){
                    vector<int> triple;
                    triple.push_back(nums[i]);
                    triple.push_back(nums[j]);
                    triple.push_back(nums[k]);
                    res.push_back(triple);
                    j++;
                    k--;
                }else if (sum < 0){
                    j++;
                }else{
                    k--;
                }
            }
        }
        
        return res;
    }
};
